### PR TITLE
Fix for connection status callback (reason IOTHUB_CLIENT_CONNECTION_NO_NETWORK) with device multiplexing

### PR DESCRIPTION
**Pull request by ewertons**
# Checklist
- [x] I have read the [contribution guidelines] (https://github.com/Azure/azure-iot-sdk-c/blob/master/.github/CONTRIBUTING.md).
- [x] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- If this is a modification that impacts the behavior of a public API
  - [ ] I edited the corresponding document in the `devdoc` folder and added or modified requirements.
- I submitted this PR against the correct branch: 
  - [x] This pull-request is submitted against the `master` branch. 
  - [x] I have merged the latest `master` branch prior to submission and re-merged as needed after I took any feedback.
  - [ ] I have squashed my changes into one with a clear description of the change.

# Reference/Link to the issue solved with this PR (if any)

https://github.com/Azure/azure-iot-sdk-c/issues/1434

# Description of the problem

Please see the github issue for details.

# Description of the solution
 

Issue is in the way that connection status callbacks are propagated by the AMQP transport.

static void on_amqp_connection_state_changed(const void* context, AMQP_CONNECTION_STATE previous_state, AMQP_CONNECTION_STATE new_state)
{
    if (context != NULL && new_state != previous_state)
    {
        AMQP_TRANSPORT_INSTANCE* transport_instance = (AMQP_TRANSPORT_INSTANCE*)context;

        // Codes_SRS_IOTHUBTRANSPORT_AMQP_COMMON_09_059: [`new_state` shall be saved in to the transport instance]
        transport_instance->amqp_connection_state = new_state;

        // Codes_SRS_IOTHUBTRANSPORT_AMQP_COMMON_09_060: [If `new_state` is AMQP_CONNECTION_STATE_ERROR, the connection shall be flagged as faulty (so the connection retry logic can be triggered)]
        if (new_state == AMQP_CONNECTION_STATE_ERROR)
        {
            LogError("Transport received an ERROR from the amqp_connection (state changed %s -> %s); it will be flagged for connection retry.", MU_ENUM_TO_STRING(AMQP_CONNECTION_STATE, previous_state), MU_ENUM_TO_STRING(AMQP_CONNECTION_STATE, new_state));
            transport_instance->transport_callbacks.connection_status_cb(IOTHUB_CLIENT_CONNECTION_UNAUTHENTICATED, IOTHUB_CLIENT_CONNECTION_NO_NETWORK, transport_instance->transport_ctx); // LINE 779.
            update_state(transport_instance, AMQP_TRANSPORT_STATE_RECONNECTION_REQUIRED);
        }

It invokes the callback only once, passing `transport_instance->transport_ctx`.
transport_ctx gets set by IoTHubTransport_SetCallbackContext (in AMQP transport, IoTHubTransport_AMQP_SetCallbackContext),
which is called by initialize_iothub_client in iothub_client_core_ll.c if the client is multiplexed.
Since the transport holds just one instance of transport_ctx, the last multiplexed client to be created (with _CreateWithTransport()) gets to be the one that gets callbacks invoked.
Notice that that is the only connection status callback that is invoked using transport_instance->transport_ctx, all others use registered_device->transport_ctx.
MQTT transport is not a problem for now since it does not support multiplexing.

### CODE REVIEW

**Comment by markrad** [comment]
I appreciate you getting to this. One comment though, it would be nice if the connection status callback function could determine if this is a single device going offline or a complete connection failure. I don't think that is possible at the moment. I know this won't get done any time soon but just some food for thought for a later iteration.

File: iothub_client/src/iothubtransport_amqp_common.c [hunk 1]
```diff
@@ -775,8 +793,19 @@ static void on_amqp_connection_state_changed(const void* context, AMQP_CONNECTIO
         // Codes_SRS_IOTHUBTRANSPORT_AMQP_COMMON_09_060: [If `new_state` is AMQP_CONNECTION_STATE_ERROR, the connection shall be flagged as faulty (so the connection retry logic can be triggered)]
         if (new_state == AMQP_CONNECTION_STATE_ERROR)
         {
+            MUX_CONN_STATUS_CB_ARGS args;
+            args.status = IOTHUB_CLIENT_CONNECTION_UNAUTHENTICATED; 
+            args.reason = IOTHUB_CLIENT_CONNECTION_NO_NETWORK;
+
+            printf("MUX_CONN_STATUS_CB_ARGS %p, %d, %d\r\n", &args, args.status, args.reason);
```
[the remark this answers is not in the document]
Comment (ewertons) [reply iothub_client/src/iothubtransport_amqp_common.c]: OMG. Should have removed those.

File: iothub_client/src/iothubtransport_amqp_common.c [hunk 2]
```diff
@@ -775,8 +793,19 @@ static void on_amqp_connection_state_changed(const void* context, AMQP_CONNECTIO
...
+
             LogError("Transport received an ERROR from the amqp_connection (state changed %s -> %s); it will be flagged for connection retry.", MU_ENUM_TO_STRING(AMQP_CONNECTION_STATE, previous_state), MU_ENUM_TO_STRING(AMQP_CONNECTION_STATE, new_state));
-            transport_instance->transport_callbacks.connection_status_cb(IOTHUB_CLIENT_CONNECTION_UNAUTHENTICATED, IOTHUB_CLIENT_CONNECTION_NO_NETWORK, transport_instance->transport_ctx);
+            
+            (void)singlylinkedlist_foreach(
```
[the remark this answers is not in the document]
Comment (ewertons) [reply iothub_client/src/iothubtransport_amqp_common.c]: That was my line of thought as well... It was always supposed to be like that. And I think few people use C SDK for multiplexing, so hopefully the impact will be small.

File: iothub_client/src/iothubtransport_amqp_common.c [hunk 3]
```diff
@@ -763,6 +763,24 @@ static int get_new_underlying_io_transport(AMQP_TRANSPORT_INSTANCE* transport_in
...
+    IOTHUB_CLIENT_CONNECTION_STATUS status;
+    IOTHUB_CLIENT_CONNECTION_STATUS_REASON reason;
+} MUX_CONN_STATUS_CB_ARGS;
+
+static void report_status_connection_callback_for_multiplexed_devices(const void* item, const void* action_context, bool* continue_processing)
+{
+    AMQP_TRANSPORT_DEVICE_INSTANCE* registered_device = (AMQP_TRANSPORT_DEVICE_INSTANCE*)item;
+    MUX_CONN_STATUS_CB_ARGS* conn_status_args = (MUX_CONN_STATUS_CB_ARGS*)action_context; 
+
+    printf("MUX_CONN_STATUS_CB_ARGS %p, %d, %d\r\n", conn_status_args, conn_status_args->status, conn_status_args->reason);
```
[the remark this answers is not in the document]
Comment (ewertons) [reply iothub_client/src/iothubtransport_amqp_common.c]: Must be removed.

File: iothub_client/src/iothubtransport_amqp_common.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (ewertons) [reply iothub_client/src/iothubtransport_amqp_common.c]: Done.

File: iothub_client/src/iothubtransport_amqp_common.c [hunk 4]
```diff
@@ -775,8 +791,17 @@ static void on_amqp_connection_state_changed(const void* context, AMQP_CONNECTIO
...
         if (new_state == AMQP_CONNECTION_STATE_ERROR)
         {
+            MUX_CONN_STATUS_CB_ARGS args;
+            args.status = IOTHUB_CLIENT_CONNECTION_UNAUTHENTICATED; 
+            args.reason = IOTHUB_CLIENT_CONNECTION_NO_NETWORK;
+
             LogError("Transport received an ERROR from the amqp_connection (state changed %s -> %s); it will be flagged for connection retry.", MU_ENUM_TO_STRING(AMQP_CONNECTION_STATE, previous_state), MU_ENUM_TO_STRING(AMQP_CONNECTION_STATE, new_state));
-            transport_instance->transport_callbacks.connection_status_cb(IOTHUB_CLIENT_CONNECTION_UNAUTHENTICATED, IOTHUB_CLIENT_CONNECTION_NO_NETWORK, transport_instance->transport_ctx);
+            
+            (void)singlylinkedlist_foreach(
```
Comment (jspaith) [inline iothub_client/src/iothubtransport_amqp_common.c]: >singlylinkedlist_foreach [](start = 18, length = 24)

I'm not that familiar with this code, but... why are we only signaling this on errors?  Shouldn't caller get a callback on the network being available also?

Comment (ewertons) [reply iothub_client/src/iothubtransport_amqp_common.c]: Hey John, it does get invoked for connection re-established. Please look here: https://github.com/Azure/azure-iot-sdk-c/blob/2020-01-22/iothub_client/src/iothubtransport_amqp_common.c#L293

Comment (jspaith) [reply iothub_client/src/iothubtransport_amqp_common.c]: OK - thanks for followup and offline conversation.  Flipping to approved.

**Pull request closed without merging**